File: libs/logger/src/bindings/zig/logger.zig
const bindings = @import("bindings.zig");

// ── Public Logger struct ─────────────────────────────────────────────

pub const Logger = struct {
    handle: c_int,
    available: bool,

    /// Create a new logger instance. If liblogger is not available, returns
    /// a no-op logger (available=false).
    ///
    /// - system: service name (e.g. "port-hook")
    /// - subsystem: sub-component (pass "" for none)
    /// - level: "debug", "info", "warn", "error"
    /// - format: "dev" (colored) or "prod" (JSON)
    pub fn init(system: [*:0]const u8, subsystem: [*:0]const u8, level: [*:0]const u8, format: [*:0]const u8) Logger {
        const syms = bindings.loadBindings() orelse return .{ .handle = 0, .available = false };
        return .{
            .handle = syms.new(system, subsystem, level, format),
            .available = true,
        };
    }

    pub fn info(self: *const Logger, msg: [*:0]const u8, context: [*:0]const u8) void {
        const syms = bindings.resolved orelse return;
        if (self.available) syms.info(self.handle, msg, context);
    }

    pub fn warn(self: *const Logger, msg: [*:0]const u8, context: [*:0]const u8) void {
        const syms = bindings.resolved orelse return;
        if (self.available) syms.warn(self.handle, msg, context);
    }

    pub fn err(self: *const Logger, msg: [*:0]const u8, context: [*:0]const u8) void {
        const syms = bindings.resolved orelse return;
        if (self.available) syms.err(self.handle, msg, context);
    }

    pub fn debug(self: *const Logger, msg: [*:0]const u8, context: [*:0]const u8) void {
        const syms = bindings.resolved orelse return;
        if (self.available) syms.debug(self.handle, msg, context);
    }

    pub fn success(self: *const Logger, msg: [*:0]const u8, context: [*:0]const u8) void {
        const syms = bindings.resolved orelse return;
        if (self.available) syms.success(self.handle, msg, context);
    }

    pub fn started(self: *const Logger, msg: [*:0]const u8, context: [*:0]const u8) void {
        const syms = bindings.resolved orelse return;
        if (self.available) syms.started(self.handle, msg, context);
    }

    pub fn plain(self: *const Logger, msg: [*:0]const u8, context: [*:0]const u8) void {
        const syms = bindings.resolved orelse return;
        if (self.available) syms.plain(self.handle, msg, context);
    }

    pub fn deinit(self: *const Logger) void {
        if (!self.available) return;
        const syms = bindings.resolved orelse return;
        syms.free(self.handle);
    }

    /// Configure the Go logger to POST entries to a server endpoint.
    pub fn setSink(server_url: [*:0]const u8, project_name: [*:0]const u8) void {
        const syms = bindings.resolved orelse return;
        syms.set_sink(server_url, project_name);
    }

    /// Flush remaining entries and stop the sink.
    pub fn closeSink() void {
        const syms = bindings.resolved orelse return;
        syms.close_sink();
    }

    /// Returns true if the Go logger library was successfully loaded.
    pub fn isAvailable(self: *const Logger) bool {
        return self.available;
    }
};

File: libs/logger/src/bindings/python/logger.py
import atexit
import json
import os
from typing import Any, Literal, Optional

try:
    from .bindings import lib
except ImportError:
    from bindings import lib

LogLevel = Literal["debug", "info", "warn", "error"]
LogFormat = Literal["dev", "prod"]


class Logger:
    """Logger that calls Go shared library via ctypes."""

    def __init__(
        self,
        system: str,
        subsystem: Optional[str] = None,
        level: Optional[LogLevel] = None,
        format: Optional[LogFormat] = None,
    ):
        self._system = system
        self._subsystem = subsystem
        self._level = level
        self._format = format

        env_level = os.getenv("LOG_LEVEL") or level or "info"
        env_format = format or ("prod" if os.getenv("NODE_ENV") == "production" else "dev")

        self.handle = lib.LoggerNew(
            system.encode("utf-8"),
            (subsystem or "").encode("utf-8"),
            env_level.encode("utf-8"),
            env_format.encode("utf-8"),
        )

        # Register cleanup
        atexit.register(self._cleanup)

    def __getattr__(self, name: str) -> "Logger":
        """
        Create a subsystem logger on-the-fly.

        Example:
            logger.nats.info("Message")  # Creates AI | NATS logger
            logger.db.error("Error")     # Creates AI | DB logger
        """
        # Avoid infinite recursion - only intercept subsystem names
        # Don't intercept private attributes, methods, or properties set in __init__
        if (name.startswith('_') or
            name in ('handle', 'info', 'warn', 'error', 'debug', 'success', 'started', 'plain', '_cleanup')):
            raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")

        # Create a new logger with the subsystem
        return Logger(self._system, name.upper(), self._level, self._format)

    def _build_context(self, domain: str, context: dict[str, Any]) -> str:
        merged = {"domain": domain, **context}
        return json.dumps(merged)

    def info(self, message: str, *, domain: str, **context: Any) -> None:
        """Log info message. ``domain`` is required to namespace logs by feature area."""
        lib.LoggerInfo(
            self.handle,
            message.encode("utf-8"),
            self._build_context(domain, context).encode("utf-8"),
        )

    def warn(self, message: str, *, domain: str, **context: Any) -> None:
        """Log warning message. ``domain`` is required to namespace logs by feature area."""
        lib.LoggerWarn(
            self.handle,
            message.encode("utf-8"),
            self._build_context(domain, context).encode("utf-8"),
        )

    def error(self, message: str, *, domain: str, **context: Any) -> None:
        """Log error message. ``domain`` is required to namespace logs by feature area."""
        lib.LoggerError(
            self.handle,
            message.encode("utf-8"),
            self._build_context(domain, context).encode("utf-8"),
        )

    def debug(self, message: str, *, domain: str, **context: Any) -> None:
        """Log debug message. ``domain`` is required to namespace logs by feature area."""
        lib.LoggerDebug(
            self.handle,
            message.encode("utf-8"),
            self._build_context(domain, context).encode("utf-8"),
        )

    def success(self, message: str, *, domain: str, **context: Any) -> None:
        """Log success message (green bullet, INFO level, status: success). ``domain`` is required."""
        ctx = {**context, "status": "success"}
        lib.LoggerSuccess(
            self.handle,
            message.encode("utf-8"),
            self._build_context(domain, ctx).encode("utf-8"),
        )

    def started(self, message: str, *, domain: str, **context: Any) -> None:
        """Log started message (INFO level, status: started). ``domain`` is required."""
        ctx = {**context, "status": "started"}
        lib.LoggerStarted(
            self.handle,
            message.encode("utf-8"),
            self._build_context(domain, ctx).encode("utf-8"),
        )

    def plain(self, message: str, *, domain: str, **context: Any) -> None:
        """Log plain message (no prefix, no color, INFO level). ``domain`` is required."""
        lib.LoggerPlain(
            self.handle,
            message.encode("utf-8"),
            self._build_context(domain, context).encode("utf-8"),
        )

    @staticmethod
    def set_sink(server_url: str, project_name: str) -> None:
        """Configure the Go logger to POST entries to a server endpoint."""
        lib.LoggerSetSink(server_url.encode("utf-8"), project_name.encode("utf-8"))

    @staticmethod
    def close_sink() -> None:
        """Flush remaining entries and stop the sink."""
        lib.LoggerCloseSink()

    def _cleanup(self) -> None:
        """Cleanup logger on exit."""
        if self.handle:
            lib.LoggerFree(self.handle)
